RecommenderSystems/pnn/README.md

```diff
@@ -0,0 +1,169 @@
+| table_size_array           | embedding table size array for sparse fields                 | *Required Argument*      |
+| persistent_path            | path for persistent kv store of embedding                    | *Required Argument*      |
+| store_type                 | OneEmbeddig persistent kv store type: `device_mem`, `cached_host_mem` or `cached_ssd` | `cached_host_mem`        |
+| cache_memory_budget_mb     | size of cache memory budget on each device in megabytes when `store_type` is `cached_host_mem` or `cached_ssd` | 1024                     |
+| amp                        | enable Automatic Mixed Precision(AMP) training or not        | False                    |
+| loss_scale_policy          | loss scale policy for AMP training: `static` or `dynamic`    | `static`                 |
+| disable_early_stop         | disable early stop or not                                    | False                    |
+| save_best_model            | save best model or not                                       | False                    |
+#### Early Stop Schema
+
+The model is evaluated at the end of every epoch. At the end of each epoch, if the early stopping criterion is met, the training process will be stopped. 
+
+The monitor used for the early stop is `val_auc - val_log_loss`. The mode of the early stop is `max`. You could tune `patience` and `min_delta` as needed.
+
+If you want to disable early stopping, simply add `--disable_early_stop` in the [train_pnn.sh](https://github.com/Oneflow-Inc/models/blob/dev_ipnn_only/RecommenderSystems/pnn/train_pnn.sh).
+
+## Getting Started
+
+A hands-on guide to train a PNN model.
+
+### Environment
+
+1.   Install OneFlow by following the steps in [OneFlow Installation Guide](https://github.com/Oneflow-Inc/oneflow#install-oneflow) or use the command line below.
+
+     ```shell
+     python3 -m pip install --pre oneflow -f https://staging.oneflow.info/branch/master/cu102
+     ```
+
+2.   Install all other dependencies listed below.
+
+```json
+CUDA: 10.2
+python: 3.8.4
+oneflow: 0.8.0
+numpy: 1.19.2
+scipy: 1.7.3
+sklearn: 1.0.2
+psutil: 5.9.0
+petastorm: 0.11.4
+pandas: 1.4.1
+pyspark: 3.2.1
+```
+
+### Dataset
+
+**Note**: 
+
+According to [the PNN paper](https://arxiv.org/pdf/1611.00144.pdf), we treat both categorical and continuous features as sparse features. 
+
+>   χ may include categorical fields (e.g., gender, location) and continuous fields (e.g., age). Each categorical field is represented as a vector of one-hot encoding, and each continuous field is represented as the value itself, or a vector of one-hot encoding after discretization. 
+
+1.   Download the [Criteo Kaggle dataset](https://www.kaggle.com/c/criteo-display-ad-challenge) and then split it using [split_criteo_kaggle.py](https://github.com/Oneflow-Inc/models/blob/dev_ipnn_only/RecommenderSystems/pnn/tools/split_criteo_kaggle.py).
```

Comment:
https://github.com/Oneflow-Inc/models/blob/dev_deepfm_spark_split/RecommenderSystems/deepfm/tools/split_criteo_kaggle.scala
split_criteo_kaggle.py is not available